### ISSUE TITLE
Activating SublimeHaskell only when .hs files are opened

### ISSUE THREAD

**Issue by jaxuru** [SublimeHaskell/SublimeHaskell#274]
Feature Request: One of the reasons SublimeText is so nice to use is how lightweight it is, supplying only what you need. Because of this, it feels a bit clunky to open sublime for something simple and non-haskell-related, like throwing together a script or something, and seeing that huge usage spike as Sublime Haskell inspects the database. It would be nice to have a preference option for ensuring Sublime Haskell only runs when a .hs file, or a folder/project with a .cabal file, is opened. I find myself disabling Sublime Haskell and only turning it on when I need it, which is awkward.

**Comment by YLiLarry** [comment]
Second this. I find that Sublime Haskell keeps reporting "Sublime Haskell: hsdev ping: no pong" in the console even when I'm writing C code. Is there a way to only enable it for Haskell code?

**Comment by iamrecursion** [comment]
I've been manually disabling and enabling it, but it's frightfully clunky to do so.

**Comment by jonaprieto** [comment]
Yes!, Sublime turns so slow, and in at some point, I'd closed it after disabling the package, and everything comes normally when I open again the next time.

**Comment by mvoidex** [comment]
Sry guys, have no time
I agree, it is annoying

**Comment by bscottm** [comment]
@jaxuru, @YLiLarry, @iamrecursion, @jonaprieto: I'll look into it in the near future.

**Comment by iamrecursion** [comment]
Excellent to hear! Your half-open pipe fix has fixed a not just one (as I said on Haskell-Cafe), but two of the issues I've been having.

**Comment by bscottm** [comment]
@iamrecursion: Thanks for the hat tip! It's going to take a bit of refactoring. Demand loading is actually implemented, but the code needs to be moved around. ST loads the plugin source in alphabetical order and the current demand loading is embedded in `autocomplete.py` (one of the first files loaded.)

**Comment by iamrecursion** [comment]
Looking forward to it! I'd offer to help if I had more free time, but not at the moment. I hope it goes smoothly for you!

**Comment by xldenis** [comment]
This would be nice, `hsdev` currently takes up 500mb as soon as I open sublime, even if I'm just writing text or editing a different language. Ensuring that `SublimeHaskell` no-ops until a haskell file loads would be greatly appreciated.

**Comment by bscottm** [comment]
@xldenis: `hsdev` should not get fired up until you load Haskell source into ST3. So, if you have a `.hs` file open, SublimeHaskell will fire up `hsdev`.

Let me get past the current release (and refactoring issues). I was working on startup refactoring, but then this whole problem of builds "hanging" and other minor issues came up. :-)

**Comment by xldenis** [comment]
![image](https://cloud.githubusercontent.com/assets/451373/24559823/be7c73d4-1605-11e7-83cb-652d945ec4f1.png)

`hsdev` booted up on sublime start, even when I only have ruby code open.

**Comment by bscottm** [comment]
@jaxuru, @YLiLarry, @iamrecursion, @jonaprieto, @mvoidex, @dabro:

Quick progress update: Getting very close to having this enhancement working, probably within the next two weeks or so. I had to refactor backend management in SublimeHaskell, which, as you can imagine, is non-trivial.

I'm at a point at which beta testing will be useful, if you're willing the help and you use `hsdev` as your backend. There is also a `none` backend, which should be self-explanatory. You will also have to change your ST user preferences -- `enable_hsdev`, `enable_ghc_mod` and related preferences are now gone, replaced with:

```yaml
    "backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567
            }
        },
        "Remote hsdev@localhost:4567": {
            "default": false,
            "backend": "hsdev",
            "options": {
                "local": false,
                "host": "localhost",
                "port": 4567
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    },
```

I do need to write the code so that you can change backends "on-the-fly" (e.g., stop `hsdev` and choose the `none` backend.)

**Comment by iamrecursion** [comment]
@bscottm Is this ready to mess with in HEAD? If so, what's the current expected behaviour?

**Comment by bscottm** [comment]
@iamrecursion: I'm a little leery of merging the `backend_refactor` branch into `HEAD` just yet -- I've done my testing, lazy loading seems to work for me. But your usage patterns and workflow are different from mine. So, instead of merging it into `HEAD`, I'd really prefer if a couple of issue reports could beta test it for me. This means working from my fork and `backend_refactor` branch, temporarily.

It'd be nice to work off a smaller issue backlog as opposed to doing a Klingon release (i.e., "Today is a good day to die! Let's release!")

**Comment by bscottm** [comment]
@iamrecursion: And I still need to migrate ghc-mod into the new backend management structure, prior to release. I'm going to defer on `hdevtools` for a while because I suspect that no one really uses it.

**Comment by bscottm** [comment]
@jaxuru, @YLiLarry, @jonaprieto, @mvoidex, @dabro: Mentioning you all so that you have visibility on the issue thread.

**Comment by iamrecursion** [comment]
@bscottm I suspect my timeline for getting to play with this might be a little longer than you're looking for, but if you're still in need of testing for this in a couple of weeks I'll be happy to help out. I hope someone else can step up to fill the plate sooner! I'm very excited to see this working.

**Comment by jaxuru** [comment]
I tested it out and first got this message when I tried to save an `.hs` file:
<img width="415" alt="screen 2017-05-16 at 17 22 28" src="https://cloud.githubusercontent.com/assets/8195503/26129358/e645b160-3a5d-11e7-9198-1cca590592bd.png">

I replaced my settings file with the `backends` preference block you included, and the pop-up didn't reappear, but hsdev still launched without any `.hs` files open and ran up a huge amount of CPU usage. Also, when I saved an `.hs` file, there was a `Checking and Linting:` message in the status bar that never disappeared.

This could potentially be an issue with how I've set up, I'm not sure if I have the right preferences. I cloned the `backend_refactor` branch of `bscottm/SublimeHaskell` into my packages folder, and my user preferences currently look like this:

```
{
"backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567
            }
        },
        "Remote hsdev@localhost:4567": {
            "default": false,
            "backend": "hsdev",
            "options": {
                "local": false,
                "host": "localhost",
                "port": 4567
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    }
}
```

**Comment by bscottm** [comment]
@jaxru: Thanks for testing. If you checkout the `master` branch, you'll get back to version 1.2.5. I'll look into the `check_lint.run_chain` problem. I did find a few places where I could defer starting up the backend until as late as possible, so things should get better on the next `git pull`.

**Comment by bscottm** [comment]
@jaxuru: If you see this in the ST console window, the issue is in `hsdev`, more specifically, in `hslint-2.0.x`:

```
<hsdev stderr> 2017-05-16 17:00:58 -0700	ERROR	:127.0.0.1:57471/127.0.0.1:57471/req/22> Scope leaves with exception: io exception: <...>\cabal\x86_64-windows-ghc-8.0.1\hlint-2.0.6\HLint.hs: openFile: does not exist (No such file or directory)
<hsdev stderr> CallStack (from HasCallStack):
<hsdev stderr>   onError, called at src\System\Log\Simple\Monad.hs:104:36 in simple-log-0.8.5-782Ab8dLFzfE0wVaBRPWR8:System.Log.Simple.Monad
```

**Comment by bscottm** [comment]
@jaxuru: Should be safe to pull an update on the `backend_refactor` branch.

**Comment by jaxuru** [comment]
I pulled from `backend_refactor` and tried again, still starting `hsdev` automatically but now the message that it is stuck on is `SublimeHaskell: Inspecting...` instead of `Checking and Linting` . The only error messages in the console are the following:

```
<hsdev stdout> 16/05/17 21:10:23 -0400	ERROR	<socket: 30>/<socket: 30>/1/update/cabal/task/package-db/scan> Error reading cache: IO exception while reading cache from /[filepath]/packages/Cache/SublimeHaskell/hsdev/cabal/user.json
```
and this:
```
<hsdev stdout> 16/05/17 21:12:00 -0400	WARN	<socket: 30>/<socket: 30>> Invalid request {"projects":[],"files":[{"file":"/[filepath]/test.hs","contents":null}],"id":"2","command":"scan","ghc-opts":[],"sandboxes":[],"paths":[],"infer":false,"docs":false,"cabal":false,"no-file":true}
Sublime Haskell: hsdev module returns error: module not inspected
```

I'm wondering if I should be doing something to reset things between trying it out, should I delete my preference files and reinstall the package to get default settings?

**Comment by bscottm** [comment]
@jaxuru: Using Package Manager to delete the plugin and recloning from git is a good idea, since it would kill off all of the older default plugin settings. I also fixed an issue with the null backend, which might be the source of your other issue (another pull... :-)

**Comment by jaxuru** [comment]
No more error pop-ups, but I'm still getting the `IO exception while reading cache` error from earlier, along with hsdev starting up without a `.hs` file, and the infinite `Checking and Linting` message. I've noticed that it actually only happens every other time I open Sublime— could it have something to do with that cache file being fixed when the error message appears, and then broken again during another start-up?

I also opened activity monitor and noticed that I had two `hsdev` processes, and when I opened Sublime I had three. I quit the processes and opened Sublime another couple of times, and had two processes again.

Also, each time that Sublime starts and doesn't start hsdev, when I open a `.hs` file, `hsdev` doesn't start, and trying to save the `.hs` file results in an error pop-up. `hsdev`'s startup seems to be only at start-up, currently.

**Comment by bscottm** [comment]
@jaxuru: `IO exception...` is generated by `hsdev` itself -- it's relatively harmless (usually while reading `userdb.) Not somethingI can fix.

Ok -- there's a new issue on how to keep SH from starting multiple `hsdev`s. That I can fix. I did find a bug where SH prematurely started a `hsdev`, but it doesn't fix the ability of SH to spawn multiples. I'll try to fix that on the next pull.

The infinite `Checking and Linting`: Does the console have a message about `Hlint.hs` not being found?

**Comment by bscottm** [comment]
@jaxuru: Sorry for the delay, go ahead and pull on the `backend_refactor` branch. Should be a little less quirky.

**Comment by bscottm** [comment]
@jaxuru: You might want to disable auto-linting -- the infinite hang is not one I can easily fix (it's a `hsdev` issue and SublimeHaskell does not receive an error response.)

**Comment by jaxuru** [comment]
Hello, sorry for the delay (I just graduated from college and flew home). Deleted the plugin and cloned again, and the multiple `hsdev` issue is fixed! It appears in my activity monitor when ST opens and disappears shortly after ST is closed.

`hsdev` still opens every time ST is opened, with a spike in usage when opened and closed. At this point, the usage is spike is very brief, so this issue is basically cosmetic (`SublimeHaskell: Inspecting user-db` appearing in the status bar) but still evidence of something not quite working correctly.

I'm wiping and reinstalling Mac Os (unrelated, just for periodic cleaning) so as soon as I reinstall SublimeText, I can see if this issue is specific to a weird installation quirk.

**Comment by bscottm** [comment]
@jaxuru: That's really weird, because when I open ST with a window that has Haskell source files, nothing happens until I switch to that window and click on the file with Haskell source. If it still happens after you reinstall, can you step me through how you started up ST, how many ST windows are open, what's in view?

**Comment by bscottm** [comment]
@jaxuru: And, thanks again for being the guinea pig. It's a real help.

**Comment by iamrecursion** [comment]
What's the status on this feature? I'm wondering regarding #318 as that depends on this work.

**Comment by bscottm** [comment]
@iamrecursion: Work-in-progress still. Two things that need fixing before release: (1) disentangle `hsinspect` from two places in the code (if you're using `ghc-mod`, you're won't have `hsinspect` installed, but you will want the equivalent functionality), (2) fix a bug that Alexandr reported during testing.

Sorry for the pause. Been crushed by an online course I'm taking. Lots of reading and paper writing.

**Comment by iamrecursion** [comment]
No problem! Thanks for the update!!

**Comment by bscottm** [comment]
@iamrecursion: Here's the problem with `hsinspect` in a nutshell -- it's a nice utility to get info out of the `.cabal` files, such as `hs-src-dirs`. Which is great, unless you're using `ghc-mod` or SublimeHaskell is just willing to support one, and only one, backend. I'm not sure the one-and-only-one backend is necessarily the right thing to do. However, parsing `.cabal` files is a real _b***h_, which I only appreciated recently (indentation-aware parsers are not trivial.)

**Comment by iamrecursion** [comment]
I understand the hope to keep supporting `ghc-mod`, but at some point don't the technical demands to provide all the features with all the backends become too much?

**Comment by bscottm** [comment]
Update: I'm close to ready to release. Sorry for the summer hiatus -- a minor professional/career event occurred that required me to do a bunch of distance learning courses (*).

Anyone opposed to a release next week? :-)

**Comment by bscottm** [comment]
For reference, the plugin's default `backends` setting looks like:

    "backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567
            }
        },
        "ghc-mod": {
            "default": false,
            "backend": "ghc-mod"
        },
        "Remote hsdev@localhost:4567": {
            "default": false,
            "backend": "hsdev",
            "options": {
                "local": false,
                "host": "localhost",
                "port": 4567
            }
        },
        "Cabal-sandboxed local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "exec-with": "cabal",
              "install-dir": "$HOME/play/haskell/hsdev"
            }
        },
        "Stack-exec local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "stack",
              "install-dir": "$HOME/play/haskell/hsdev"
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    }

**Backend options**: The `ghc-mod` and `none` backends do not have additional options and ignore whatever options are supplied. The `hsdev` backend supports the following options:

    port:        [default: "4567"]      TCP port where SublimeHaskell will connect to hsdev
    host:        [defualt: "localhost"] Host name SublimeHaskell will connect to hsdev
    local:       [default: true]        Should SublimeHaskell create a local hsdev server process? Specify false
                                        if you want to connect to a remote hsdev server
    exec-with:   [no default]           Should SublimeHaskell use "stack" or "cabal" to when starting the hsdev
                                        executable? Valid options are "stack" and "cabal".
    install-dir: [no default]           Used in conjunction with "exec-with"; this is the directory where
                                        SublimeHaskell go to "cabal exec" or "stack exec" the hsdev executable.

**Comment by bscottm** [comment]
@iamrecursion, @jaxuru, @YLiLarry, @jonaprieto: 2.0.0 should now be available on Package Control and this new feature should be implemented.

**Comment by jonaprieto** [comment]
@bscottm Great to read this!. What is the commit for closing this issue?

**Comment by jonaprieto** [comment]
>  It wasn't just a single commit.

I know, of course, but that's was my point. Without referencing this issue (i.e., just including `#274`) in each commit to solve it, it's pretty hard to follow the changes. 

>  If you're really interested, you could diff between the 1.2.5 and 2.0.0 tags.

 I'm but I will not do that for sure, going from a minor version to a major version, and also regarding of ~95 commits, maybe on vacations.

**Comment by jaxuru** [comment]
Thank you so much! I'll test it out later today. Sorry I disappeared during testing, I moved to a new city.

**Comment by bscottm** [comment]
@jaxru: I can appreciate.

**Comment by bscottm** [comment]
@jonaprieto: I could have done a squash commit. Even then, there was a lot of code that changed. Getting this to work was non-trivial.

**Comment by jonaprieto** [comment]
Thanks anyway for dedicating so much time. Btw, I'm using now the plugin, and it's working well so far.

**Comment by bscottm** [comment]
@jonaprieto: Thanks for the thanks! FWIW -- there were ~5600 additions, ~4500 deletions. So, a lot of modified code. There are still a few bugs here and there, as there are and always will be. Just finished unbreaking autocompletion... one of those critical features.
